File: go/129.go

package main


// 6 star, 典型的dfs, curSum每次要乘以10然后加上当前节点的值，如果遇到叶节点，将当前的值加到总和中
func sumNumbers(root *TreeNode) int {
	if root == nil {return 0}
	sum := 0
	sum = dfs(root, 0, sum)
	return sum

}


func dfs(node *TreeNode, curSum int, sum int) int {
	if node.Left == nil && node.Right == nil {
		curSum = curSum * 10 + node.Val
		sum += curSum
		return sum
	}
	if node.Left != nil {
		sum = dfs(node.Left, curSum*10+node.Val, sum)
	}
	if node.Right != nil {
		sum = dfs(node.Right, curSum*10+node.Val, sum)
	}
	return sum
	}

File: v2/129.py






# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # 5 star, dfs, 注意这里使用了列表类型的参数来记录所有根到叶子组成的数字，因为列表是可变对象
        if not root:
            return 0
        rs = []
        self.dfs(root, str(root.val), rs)
        return sum(rs)

    def dfs(self, node, path, rs):
        if not node.left and not node.right:
            rs.append(int(path))
            return
        if node.left:
            self.dfs(node.left, path + str(node.left.val), rs)
        if node.right:
            self.dfs(node.right, path + str(node.right.val), rs)

# a = TreeNode(9)
# print(Solution().sumNumbers(a))